examples/api/test/material/text_form_field/text_form_field.2_test.dart

```diff
@@ -0,0 +1,74 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+import 'package:flutter_api_samples/material/text_form_field/text_form_field.2.dart' as example;
+import 'package:flutter_test/flutter_test.dart';
+
+void main() {
+  group('TextFormFieldExample2 Widget Tests', () {
+
+    testWidgets('Input validation handles empty, incorrect, and short usernames', (WidgetTester tester) async {
+      await tester.pumpWidget(const example.TextFormFieldExampleApp());
+      final Finder textFormField = find.byType(TextFormField);
+      final Finder saveButton = find.byType(TextButton);
+
+      await tester.enterText(textFormField, '');
+      await tester.pump();
+      await tester.tap(saveButton);
+      await tester.pump();
+      expect(find.text('This field is required'), findsOneWidget);
+
+      await tester.enterText(textFormField, 'jo hn');
+      await tester.tap(saveButton);
+      await tester.pump();
+      expect(find.text('Username must not contains any spaces'), findsOneWidget);
+
+      await tester.enterText(textFormField, 'jo');
+      await tester.tap(saveButton);
+      await tester.pump();
+      expect(find.text('username should be at least 3 characters long'), findsOneWidget);
+    });
+
+    testWidgets('Async validation feedback is handled correctly', (WidgetTester tester) async {
+      await tester.pumpWidget(const example.TextFormFieldExampleApp());
+      final Finder textFormField = find.byType(TextFormField);
+      final Finder saveButton = find.byType(TextButton);
+
+      // Simulate entering a username already taken
+      await tester.enterText(textFormField, 'jack');
+      await tester.pump();
+      await tester.tap(saveButton);
+      await tester.pumpAndSettle(const Duration(seconds: 3)); // Wait for async call
```

Comment:
Nit: Not sure, but you might be able to just use `pump` instead of `pumpAndSettle` here (keeping the duration).

Reply:
Also, I would split this up. Pump one frame here,  then expect that the error message is not found (or no error message at all is found). Then `pump` the 3 second duration and continue the rest of the test as-is.